signac/contrib/_index.py

```diff
@@ -0,0 +1,294 @@
+from numbers import Number
+
+from ..errors import InvalidKeyError
+from .collection import (
+    _INDEX_OPERATORS,
+    _check_logical_operator_argument,
+    _DictPlaceholder,
+    _find_with_index_operator,
+    _float,
+    _TypedSetDefaultDict,
+)
+from .utility import _nested_dicts_to_dotted_keys, _to_hashable
+
+logger = logging.getLogger(__name__)
+
+_PRIMARY_KEY = "_id"
+
+
+class _Index(dict):
+    """A searchable collection of dicts.
+
+    The _Index class is a :class:`dict` that maps from ids to :class:`dict`s.
+    The :class:`dict`s stored as values can be searched by their contained keys
+    and values, returning ids for the values matching the provided query. The
+    query syntax is based on MongoDB, though this class does not aim to match
+    the API of MongoDB's Collection class.
+
+    The dictionary values may be nested (may contain other dicts or lists), but
+    have two restrictions. First, the data must be JSON-encodable. Second, the
+    keys in the dictionary may not contain dots (``.``).
+
+    For example, suppose we are given dictionaries of member data containing a
+    `name` key and an `age` key along with unique identifiers acting as a
+    primary key for each member. We can find the name of all members that are
+    age 32 like this:
+
+    .. code-block:: python
+
+        members = _Index({
+            '0': {'name': 'John',  'age': 32},
+            '1': {'name': 'Alice', 'age': 28},
+            '2': {'name': 'Kevin', 'age': 32},
+            # ...
+        })
+
+        for member_id in members.find({'age': 32}):
+            print(member_id)  # prints 0 and 2
+
+    Because this class inherits from :class:`dict`, it can be constructed in
+    any of the same ways as a :class:`dict`, like ``_Index(**kwargs)``,
+    ``_Index(mapping, **kwargs)``, or ``_Index(iterable, **kwargs)``.
+
+    """
+
+    def build_index(self, key):
+        """Build index for a given key.
+
+        This is a highly performance critical code path.
+
+        Parameters
+        ----------
+        key : str
+            The key on which the index is built.
+
+        Returns
+        -------
+        :class:`~_TypedSetDefaultDict`
+            Index for key.
+
+        Raises
+        ------
+        :class:`~signac.errors.InvalidKeyError`
+            The dict contains invalid keys.
+
+        """
+        logger.debug(f"Building index for key '{key}'...")
+        nodes = key.split(".")
+        index = _TypedSetDefaultDict()
+
+        for _id, doc in self.items():
+            try:
+                v = doc
+                # Recursively access nested values from dotted keys.
+                for n in nodes:
+                    v = v[n]
+            except (KeyError, TypeError):
+                pass
+            else:
+                if type(v) is list:  # Avoid isinstance for performance.
+                    index[_to_hashable(v)].add(_id)
+                elif type(v) is dict:  # Avoid isinstance for performance.
```

Comment:
```suggestion
                # `isinstance(instance, cls)` is typically faster than `type(instance) is cls`
                # when the answer is True, but it is slower when it is False. Since we
                # expect lists and dicts to occur infrequently here, we optimize for the
                # False path using the `type` based check.
                if type(v) is list:
                    index[_to_hashable(v)].add(_id)
                elif type(v) is dict:
```